exporter/exporterhelper/resource_to_label.go

```diff
@@ -0,0 +1,140 @@
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+)
+
+// ResourceToLabelSettings defines configuration for converting resource attributes to labels
+type ResourceToLabelSettings struct {
+	// Enabled indicates whether to not convert resource attributes to labels
+	Enabled bool `mapstructure:"enabled"`
+}
+
+// CreateDefaultResourceToLabelSettings returns the default settings for ResourceToLabelSettings.
+func createDefaultResourceToLabelSettings() ResourceToLabelSettings {
+	return ResourceToLabelSettings{
+		Enabled: false,
+	}
+}
+
+// convertResourceToLabels converts all resource attributes to metric labels
+func convertResourceToLabels(md pdata.Metrics) pdata.Metrics {
+	cloneMd := md.Clone()
+	rms := cloneMd.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		resource := rms.At(i).Resource()
+
+		labelMap := extractLabelsFromResource(&resource)
+
+		ilms := rms.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ilm := ilms.At(j)
+			if ilm.IsNil() {
+				continue
+			}
+			metricSlice := ilm.Metrics()
+			for k := 0; k < metricSlice.Len(); k++ {
+				metric := metricSlice.At(k)
+				if metric.IsNil() {
+					continue
+				}
+				addLabelsToMetric(&metric, labelMap)
+			}
+		}
+	}
+	return cloneMd
+}
+
+// extractAttributesFromResource extracts the attributes from a given resource and
+// returns them as a StringMap.
+func extractLabelsFromResource(resource *pdata.Resource) pdata.StringMap {
+	labelMap := pdata.NewStringMap()
+
+	attrMap := resource.Attributes()
+	attrMap.ForEach(func(k string, av pdata.AttributeValue) {
+		stringLabel := tracetranslator.AttributeValueToString(av, false)
+		labelMap.Insert(k, stringLabel)
+	})
+	return labelMap
+}
+
+// addLabelsToMetric adds additional labels to the given metric
+func addLabelsToMetric(metric *pdata.Metric, labelMap pdata.StringMap) {
+	switch metric.DataType() {
+	case pdata.MetricDataTypeIntGauge:
+		addLabelsToIntDataPoints(metric.IntGauge().DataPoints(), labelMap)
+	case pdata.MetricDataTypeDoubleGauge:
+		addLabelsToDoubleDataPoints(metric.DoubleGauge().DataPoints(), labelMap)
+	case pdata.MetricDataTypeIntSum:
+		addLabelsToIntDataPoints(metric.IntSum().DataPoints(), labelMap)
+	case pdata.MetricDataTypeDoubleSum:
+		addLabelsToDoubleDataPoints(metric.DoubleSum().DataPoints(), labelMap)
+	case pdata.MetricDataTypeIntHistogram:
+		addLabelsToIntHistogramDataPoints(metric.IntHistogram().DataPoints(), labelMap)
+	case pdata.MetricDataTypeDoubleHistogram:
+		addLabelsToDoubleHistogramDataPoints(metric.DoubleHistogram().DataPoints(), labelMap)
+	}
+}
+
+func addLabelsToIntDataPoints(ps pdata.IntDataPointSlice, newLabelMap pdata.StringMap) {
+	for i := 0; i < ps.Len(); i++ {
+		dataPoint := ps.At(i)
+		if dataPoint.IsNil() {
+			continue
+		}
+		newLabelMap.ForEach(func(k, v string) {
```

Comment:
These 3 lines are code duplicate, can have a simple func.

Reply:
Yea, they look similar but are of different types. From my understanding, for creating a different function we need to introduce some kind of interface/wrapping over these types. I think maybe it's OK to have this simple and straight forward solution rather than making it more complex.
I also found some similar patterns in our existing codebase. your call- would appreciate a second thought.

Reply:
Are they? You have two StringMap.

Reply:
Never mind. I got it wrong earlier- I thought about about line 97-100. Thanks, it makes the code nicer. Pushed an update.